### PR TITLE
nautilus: rgw: Silence warning: control reaches end of non-void function

### PR DESCRIPTION
**Pull request by epuertat**
backport tracker: https://tracker.ceph.com/issues/42836

---

backport of https://github.com/ceph/ceph/pull/28809
parent tracker: https://tracker.ceph.com/issues/40747

this backport was staged using ceph-backport.sh version 15.0.0.6950
find the latest version at https://github.com/ceph/ceph/blob/master/src/script/ceph-backport.sh

### CODE REVIEW

**Comment by smithfarm** [comment]
@epuertat We typically do not add "Signed-off-by" lines after the `(cherry picked from commit ...)` line (the first commit in this PR has two, the second none). But it's up to you whether you want to put them there or not.

**Comment by epuertat** [comment]
> @epuertat We typically do not add "Signed-off-by" lines after the `(cherry picked from commit ...)` line (the first commit in this PR has two, the second none). But it's up to you whether you want to put them there or not.

I have all git commands aliased with `-sS`, that's why the commits are signed-off and GPG-signed.

As I had to edit the first commit a couple of times to restore some (unwanted) whitespacing that my editor had removed automatically, I likely duplicated the signoffs. And for the last one, the "create-backport.sh" script doesn't use my aliases. But no problem, I'll remove them.

**Comment by epuertat** [comment]
- [Packages](https://shaman.ceph.com/builds/ceph/wip-epuertat-testing-2019-11-18-1029/) (ongoing)
- QA (pending)

**Comment by callithea** [comment]
>     * [QA](http://pulpito.ceph.com/epuertat-2019-11-18_12:48:49-rados:dashboard-wip-epuertat-testing-2019-11-18-1029-distro-basic-smithi/) (ongoing)

- https://tracker.ceph.com/issues/42708 (but this was on master plus the mentioned patch in the tracker issue on top)..

File: src/rgw/rgw_rest_s3.cc [hunk 1]
```diff
@@ -3523,18 +3523,20 @@ RGWOp *RGWHandler_REST_Service_S3::op_post()
...
-      return new RGWListBucket_ObjStore_S3v2;
-    } } else {
-    return new RGWStatBucket_ObjStore_S3;    
-  }   }
+    int list_type = 1;
+    s->info.args.get_int("list-type", &list_type, 1);
+    switch (list_type) {
+      case 1:
+        return new RGWListBucket_ObjStore_S3;
+      default:
```
Comment (cbodley) [inline src/rgw/rgw_rest_s3.cc]: this is missing two important lines from the initial pr:
```c++
      case 2:
        return new RGWListBucket_ObjStore_S3v2;
```

Comment (epuertat) [reply src/rgw/rgw_rest_s3.cc]: Thanks @cbodley ! When I started backporting this, https://github.com/ceph/ceph/pull/29849 hasn't been merged yet, so I assumed I had to remove any reference to v2. I'll amend this.

Comment (epuertat) [reply src/rgw/rgw_rest_s3.cc]: Thanks @cbodley ! When I started backporting this, https://github.com/ceph/ceph/pull/29849 hasn't been merged yet, so I assumed I had to remove any reference to v2. I'll amend this.

**Comment by epuertat** [comment]
- [Packages](https://shaman.ceph.com/builds/ceph/wip-epuertat-testing2-2019-11-18-1951/) (ongoing)
- QA

**Comment by callithea** [comment]
Hi @epuertat, I included this PR into [my QA ](http://pulpito.ceph.com/laura-2019-11-18_15:26:14-rados:dashboard-wip-laura-testing-dashboard-nautilus-batch-4-distro-basic-mira/) run and one job failed with (**note: I need to retry it with the latest commit of this PR**):
```2019-11-18T16:23:34.508 INFO:tasks.cephfs_test_runner:======================================================================
2019-11-18T16:23:34.509 INFO:tasks.cephfs_test_runner:ERROR: test_create_s3 (tasks.mgr.dashboard.test_rgw.RgwUserSubuserTest)
2019-11-18T16:23:34.509 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
2019-11-18T16:23:34.509 INFO:tasks.cephfs_test_runner:Traceback (most recent call last):
2019-11-18T16:23:34.509 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-laura-testing-dashboard-nautilus-batch-4/qa/tasks/mgr/dashboard/test_rgw.py", line 59, in setUp
2019-11-18T16:23:34.509 INFO:tasks.cephfs_test_runner:    super(RgwTestCase, self).setUp()
2019-11-18T16:23:34.509 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-laura-testing-dashboard-nautilus-batch-4/qa/tasks/mgr/dashboard/helper.py", line 154, in setUp
2019-11-18T16:23:34.509 INFO:tasks.cephfs_test_runner:    self.wait_for_health_clear(20)
2019-11-18T16:23:34.509 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-laura-testing-dashboard-nautilus-batch-4/qa/tasks/ceph_test_case.py", line 132, in wait_for_health_clear
2019-11-18T16:23:34.510 INFO:tasks.cephfs_test_runner:    self.wait_until_true(is_clear, timeout)
2019-11-18T16:23:34.510 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-laura-testing-dashboard-nautilus-batch-4/qa/tasks/ceph_test_case.py", line 164, in wait_until_true
2019-11-18T16:23:34.510 INFO:tasks.cephfs_test_runner:    raise RuntimeError("Timed out after {0}s".format(elapsed))
2019-11-18T16:23:34.510 INFO:tasks.cephfs_test_runner:RuntimeError: Timed out after 20s
2019-11-18T16:23:34.510 INFO:tasks.cephfs_test_runner:
2019-11-18T16:23:34.510 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
2019-11-18T16:23:34.510 INFO:tasks.cephfs_test_runner:Ran 82 tests in 1301.873s
2019-11-18T16:23:34.510 INFO:tasks.cephfs_test_runner:
2019-11-18T16:23:34.511 INFO:tasks.cephfs_test_runner:FAILED (errors=1)
2019-11-18T16:23:34.511 INFO:tasks.cephfs_test_runner:
2019-11-18T16:23:34.511 INFO:tasks.cephfs_test_runner:======================================================================
2019-11-18T16:23:34.511 INFO:tasks.cephfs_test_runner:ERROR: test_create_s3 (tasks.mgr.dashboard.test_rgw.RgwUserSubuserTest)
2019-11-18T16:23:34.511 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
2019-11-18T16:23:34.511 INFO:tasks.cephfs_test_runner:Traceback (most recent call last):
2019-11-18T16:23:34.511 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-laura-testing-dashboard-nautilus-batch-4/qa/tasks/mgr/dashboard/test_rgw.py", line 59, in setUp
2019-11-18T16:23:34.511 INFO:tasks.cephfs_test_runner:    super(RgwTestCase, self).setUp()
2019-11-18T16:23:34.512 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-laura-testing-dashboard-nautilus-batch-4/qa/tasks/mgr/dashboard/helper.py", line 154, in setUp
2019-11-18T16:23:34.512 INFO:tasks.cephfs_test_runner:    self.wait_for_health_clear(20)
2019-11-18T16:23:34.512 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-laura-testing-dashboard-nautilus-batch-4/qa/tasks/ceph_test_case.py", line 132, in wait_for_health_clear
2019-11-18T16:23:34.512 INFO:tasks.cephfs_test_runner:    self.wait_until_true(is_clear, timeout)
2019-11-18T16:23:34.512 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-laura-testing-dashboard-nautilus-batch-4/qa/tasks/ceph_test_case.py", line 164, in wait_until_true
2019-11-18T16:23:34.512 INFO:tasks.cephfs_test_runner:    raise RuntimeError("Timed out after {0}s".format(elapsed))
2019-11-18T16:23:34.512 INFO:tasks.cephfs_test_runner:RuntimeError: Timed out after 20s```

**Comment by smithfarm** [comment]
Note: this fixes a regression [1] introduced by https://github.com/ceph/ceph/pull/29849

[1] https://tracker.ceph.com/issues/42881

**Comment by epuertat** [comment]
```bash
2019-11-18T16:23:34.511 INFO:tasks.cephfs_test_runner:======================================================================
2019-11-18T16:23:34.511 INFO:tasks.cephfs_test_runner:ERROR: test_create_s3 (tasks.mgr.dashboard.test_rgw.RgwUserSubuserTest)
2019-11-18T16:23:34.511 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
2019-11-18T16:23:34.511 INFO:tasks.cephfs_test_runner:Traceback (most recent call last):
2019-11-18T16:23:34.511 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-laura-testing-dashboard-nautilus-batch-4/qa/tasks/mgr/dashboard/test_rgw.py", line 59, in setUp
2019-11-18T16:23:34.511 INFO:tasks.cephfs_test_runner:    super(RgwTestCase, self).setUp()
2019-11-18T16:23:34.512 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-laura-testing-dashboard-nautilus-batch-4/qa/tasks/mgr/dashboard/helper.py", line 154, in setUp
2019-11-18T16:23:34.512 INFO:tasks.cephfs_test_runner:    self.wait_for_health_clear(20)
2019-11-18T16:23:34.512 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-laura-testing-dashboard-nautilus-batch-4/qa/tasks/ceph_test_case.py", line 132, in wait_for_health_clear
2019-11-18T16:23:34.512 INFO:tasks.cephfs_test_runner:    self.wait_until_true(is_clear, timeout)
2019-11-18T16:23:34.512 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-laura-testing-dashboard-nautilus-batch-4/qa/tasks/ceph_test_case.py", line 164, in wait_until_true
2019-11-18T16:23:34.512 INFO:tasks.cephfs_test_runner:    raise RuntimeError("Timed out after {0}s".format(elapsed))
2019-11-18T16:23:34.512 INFO:tasks.cephfs_test_runner:RuntimeError: Timed out after 20s```
```

@cbodley: any clue on whether the above QA failure is a flappy one or the smell of a real issue? Thanks!

**Comment by mattbenjamin** [comment]
I think you can read radosgw's debug logs.  If it crashed you'd see evidence there.

**Comment by cbodley** [comment]
[teuthology.log](http://qa-proxy.ceph.com/teuthology/laura-2019-11-18_15:26:14-rados:dashboard-wip-laura-testing-dashboard-nautilus-batch-4-distro-basic-mira/4519481/teuthology.log):
```
2019-11-18T16:23:34.287 INFO:teuthology.orchestra.run.mira063.stdout:{"checks":{"MON_DOWN":{"severity":"HEALTH_WARN","summary":{"message":"1/3 mons down, quorum a,c"}}},"status":"HEALTH_WARN"}
2019-11-18T16:23:34.287 INFO:tasks.cephfs_test_runner:test_create_s3 (tasks.mgr.dashboard.test_rgw.RgwUserSubuserTest) ... ERROR
```
[ceph-mon.b.log.gz](http://qa-proxy.ceph.com/teuthology/laura-2019-11-18_15:26:14-rados:dashboard-wip-laura-testing-dashboard-nautilus-batch-4-distro-basic-mira/4519481/remote/mira117/log/ceph-mon.b.log.gz):
```
2019-11-18 16:22:46.512 7ff99dc15700 -1 *** Caught signal (Segmentation fault) **
 in thread 7ff99dc15700 thread_name:msgr-worker-1

 ceph version 14.2.4-796-gbe178ff (be178ffcfbf6a9853fa6eb515d0dd3862134eeb0) nautilus (stable)
 1: (()+0x12890) [0x7ff9a6013890]
 NOTE: a copy of the executable, or `objdump -rdS <executable>` is needed to interpret this.
```
the monitor crashed without a stack trace. i haven't seen this before, but it doesn't look related to radosgw

**Comment by epuertat** [comment]
Thanks for the hints @cbodley @mattbenjamin !

@callithea given the other 3 jobs of your QA run passed and this is unrelated to any Mon change, are you ok with merging this one?

**Comment by smithfarm** [comment]
@epuertat This is an rgw PR, not a dashboard PR. Ordinarily it would need to be included in an `rgw` suite integration run.

**Comment by smithfarm** [comment]
@yuriw Needed for v14.2.5 . . . can we get it in?

File: src/rgw/rgw_rest_s3.cc [hunk 2]
```diff
@@ -2662,7 +2662,7 @@ void RGWGetLC_ObjStore_S3::execute()
 void RGWGetLC_ObjStore_S3::send_response()
 {
   if (op_ret) {
-    if (op_ret == -ENOENT) {	
+    if (op_ret == -ENOENT) {
```
Comment (smithfarm) [inline src/rgw/rgw_rest_s3.cc]: this whitespace change is not in the master PR?

**Comment by epuertat** [comment]
Closed in favor of https://github.com/ceph/ceph/pull/31742

**Pull request closed without merging**